### PR TITLE
Add methods to create, update, and delete secrets

### PR DESCRIPTION
**Pull request by chaserb**
This PR addresses #4 . It adds methods to create, update, and delete secrets. Along those lines, this PR also adds accessors for secret templates, including a template's password generation function.

This PR is a pre-requisite for [issue 26](https://github.com/thycotic/terraform-provider-tss/issues/26) in [terraform-provider-tss](https://github.com/thycotic/terraform-provider-tss)

### CODE REVIEW

File: server/secret.go [hunk 1]
```diff
@@ -12,18 +12,22 @@ const resource = "secrets"
 
 // Secret represents a secret from Thycotic Secret Server
 type Secret struct {
-	Name                                string
-	FolderID, ID, SiteID                int
-	SecretTemplateID, SecretPolicyID    int
-	Active, CheckedOut, CheckOutEnabled bool
-	Fields                              []SecretField `json:"Items"`
+	Name                                                                       string        `json:",omitempty"`
+	FolderID, ID, SiteID                                                       int           `json:",omitempty"`
```
Comment (amigus) [inline server/secret.go:16]: The Name and IDs shouldn't be `omitempty`. I suggest:
```golang
type Secret struct {
	Name                                                                        string
	FolderID, ID, SiteID, SecretTemplateID, SecretPolicyID                      int
	CheckOutIntervalMinutes, LauncherConnectAsSecretID, PasswordTypeWebScriptID int           `json:",omitempty"`
	Active, CheckedOut, CheckOutEnabled                                         bool          `json:",omitempty"`
	AutoChangeEnabled, CheckOutChangePasswordEnabled, DelayIndexing             bool          `json:",omitempty"`
	EnableInheritPermissions, EnableInheritSecretPolicy, ProxyEnabled           bool          `json:",omitempty"`
	RequiresComment, SessionRecordingEnabled, WebLauncherRequiresIncognitoMode  bool          `json:",omitempty"`
	Fields                                                                      []SecretField `json:"Items,omitempty"`
}
```

File: server/secret.go [hunk 2]
```diff
@@ -12,18 +12,22 @@ const resource = "secrets"
...
+	LauncherConnectAsSecretID, PasswordTypeWebScriptID                         int           `json:",omitempty"`
+	Active, CheckedOut, CheckOutEnabled                                        bool          `json:",omitempty"`
+	AutoChangeEnabled, CheckOutChangePasswordEnabled, DelayIndexing            bool          `json:",omitempty"`
+	EnableInheritPermissions, EnableInheritSecretPolicy, ProxyEnabled          bool          `json:",omitempty"`
+	RequiresComment, SessionRecordingEnabled, WebLauncherRequiresIncognitoMode bool          `json:",omitempty"`
+	Fields                                                                     []SecretField `json:"Items,omitempty"`
 }
 
 // SecretField is an item (field) in the secret
 type SecretField struct {
```
Comment (amigus) [inline server/secret.go:27]: Analogously:
```
type SecretField struct {
	ItemID, FieldID, FileAttachmentID     int
	FieldName, Slug                       string
	FieldDescription, Filename, ItemValue string `json:",omitempty"`
	IsFile, IsNotes, IsPassword           bool   `json:",omitempty"`
}
```

File: server/secret_template.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"log"
+	"strconv"
+)
+
+// templateResource is the HTTP URL path component for the secret templates resource
+const templateResource = "secret-templates"
+
+// SecretTemplate represents a secret template from Thycotic Secret Server
+type SecretTemplate struct {
+	Name   string                `json:",omitempty"`
```
Comment (amigus) [inline server/secret_template.go:15]: I don't think any of these should be `omitempty`

File: server/secret_template.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	ID     int                   `json:",omitempty"`
+	Fields []SecretTemplateField `json:",omitempty"`
+}
+
+// SecretTemplateField is a field in the secret template
+type SecretTemplateField struct {
```
Comment (amigus) [inline server/secret_template.go:21]: I suspect most of these shouldn't be either.

**Review by amigus — CHANGES_REQUESTED** [review]
These changes look great!

The only thing is too much `omitempty`. Please update to only use `omitempty` where appropriate.

After that minor change, we will be happy to take these changes. Thanks!

File: server/secret.go (hunk 1 above)
Comment (chaserb) [reply server/secret.go]: Done. The only one I left in the `omitempty` camp is `SecretPolicyID`. The default value of `0` causes the server to complain about an invalid policy id.

File: server/secret.go (hunk 2 above)
Comment (chaserb) [reply server/secret.go:27]: Done.

File: server/secret_template.go (hunk 3 above)
Comment (chaserb) [reply server/secret_template.go]: Removed all the tags.

File: server/secret_template.go (hunk 4 above)
Comment (chaserb) [reply server/secret_template.go:21]: Removed all the tags.

**Comment by chaserb** [comment]
Hello @amigus . All the issues have been resolved. Let me know what you think.

**Comment by EndlessTrax** [comment]
Hi @chaserb, Thank you so much for this PR. It's a lot of work for which we're very grateful. 

My only question is why use `omitempty` at all? My understanding (please correct me if I'm wrong) is that it will remove the `key` from the JSON string if the value is `0` or `nil`. This means the SDK will return a different payload than the Secret Server API would, which could be potentially confusing, and will need explicit documentation. 

So I guess my question really is, what problem does `omitempty` solve that makes this tradeoff worthwhile?

**Comment by chaserb** [comment]
Hi @EndlessTrax , It looks like we can get away with `omitempty` altogether with the exception of `SecretPolicyID` and `PasswordTypeWebScriptID`.  You are correct that the `omitempty` removes the item from from JSON if either of those two values are `0`. In the case of these two IDs, however, I'm afraid that is what's required. Leaving a value of `0` for either of these two attributes causes the server to complain about invalid IDs. 

This is what I get for when leaving the `"SecretPolicyID":0` in the request body:

```
2022/01/10 12:57:00 [DEBUG] response body:
    secret_test.go:72: calling secrets.CreateSecret: 400 Bad Request: {
          "errorCode": "API_InvalidSecretPolicy (SecretPolicyId)",
          "message": "API_InvalidSecretPolicy (SecretPolicyId)"
        }
```

This is what I get when I leave `"PasswordTypeWebScriptID":0` in the request body:

```
2022/01/10 12:59:09 [DEBUG] response body:
    secret_test.go:97: calling secrets.UpdateSecret: 400 Bad Request: {
          "errorCode": "API_InvalidPasswordTypeWebScript",
          "message": "The specified password changing script is invalid or inactive."
        }
```

I've tested this both on the hosted `https://my-tenant.secretservercloud.com/` and on our on-premise v11.0 server.

I'll make another commit that pares `omitempty` down to the bare minimum.

**Comment by EndlessTrax** [comment]
Sounds good to me, thanks @chaserb. Once @amigus approves the changes, I'll merge. 

Thanks again for all your hard work.

**Review by amigus — APPROVED** [review]
Thanks for your contribution and diligence!

**Pull request merged**